packages/integration-sdk-core/src/types/step.ts

```diff
@@ -81,9 +84,21 @@ export interface GraphObjectIndexMetadata {
   enabled: boolean;
 }
 
+export interface GraphObjectSchema extends IntegrationEntitySchema {
+  $schema?: string;
+  $id?: string;
+  description?: string;
+  additionalProperties?: boolean;
```

Comment:
Do we want to make some of these `required`? It could prevent us from bugs like what we saw in `graph-aws`

Reply:
e.g. `properties`, `required`, `additionalProperties`

Reply:
```diff
export interface GraphObjectSchema extends IntegrationEntitySchema {
  $schema?: string;
  $id?: string;
  description?: string;
- additionalProperties?: boolean;
+ additionalProperties: boolean;
+ properties: Required<IntegrationEntitySchema>['properties'];
+ required: Required<IntegrationEntitySchema>['required'];
}
```

Reply:
Considering the form of https://github.com/JupiterOne/data-model/blob/178d8838d6004a45ff97b229fd8a0dafd3b50e6f/src/index.ts#L6:

```ts
export type IntegrationEntitySchema = {
  $ref?: string;
  allOf?: IntegrationEntitySchema[];
  properties?: {
    [propertyName: string]: any;
  };
  required?: string[];
};
```

Nothing is required there, which makes sense, allowing for various combinations of those properties.

For this `GraphObjectSchema`, I think we could make `description` required. I'm confused about what `$schema` does here in the context of the metadata for an entity. Will `$id` ever be provided, perhaps we could remove that? We could probably use some documentation on these properties.

What do you think about naming this `EntitySchemaExtension`?

Reply:
TBH I think it's bad that we're not using the published NPM types for JSON Schema. 

`yarn add --dev @types/json-schema`

```ts
import { JSONSchema7 } from 'json-schema'
```

Perhaps this type should _actually_ be something like the following?

```ts
type GraphObjectSchema = Required<Pick<JSONSchema7, 'additionalProperties' | 'required' | 'properties'>>;
```

Reply:
Personally I don't see why we would want to document a `description` if we're just using this to describe step entity metadata.

Reply:
Also I don't love `EntitySchemaExtension` because I would really like to see this used on relationships as well, in cases where relationships will have properties on them (example, permissions relationships or firewall rule relationships)

Reply:
Ohh, nice find on the `@types/json-schema`! I wonder if I totally overlooked that or if it's relatively new. I think it would be wise to move to use those types.

Reply:
Tracking:
- https://github.com/JupiterOne/sdk/issues/620
- https://github.com/JupiterOne/data-model/issues/147